### ISSUE TITLE
Bow

### ISSUE THREAD

**Issue by ghost** [smoku/ecto_arc#2]
I looked at the code for Bow. 

Will this work for a new created record:

**def store_dir(file) do
    "users/#{file.scope.id}/avatar"
end**

Since it is a file upload on a new record being created, I don't think **file.scope.id** would exist, it would result in a null error.

Is that right or I am misunderstanding?

**Comment by smoku** [comment]
@AppyCat yes it will work, this + parallel processing + few other things are exactly the reasons we wrote bow. Since Ecto does not have callbacks you need to call **store** function manually. 

Example code:

```
changeset = User.changeset(%User{}, params)

case Repo.insert(changeset) do
  {:ok, user} -> 
    Bow.Ecto.store!(user)
    
    render(conn, "show.json", user: user)
  {:error, changeset} ->
    render_changeset_error(conn, changeset)
end
```

Additionally you can also wrap it all with **Repo.transaction**.

Bow is quite similar to carrierwave but leverages the power of elixir in parallel uploads and processing. I encourage you to simply read the code since this library is really small :)

**Comment by smoku** [comment]
Bow was only used in our own project but recently we decided to open source it :)

**Comment by ghost** [comment]
Thank you for the update. I would say overall without proper documentation, really clear cut for Phoenix ver 1.3, it actually makes more sense to roll your own solution just using ex_aws.

In my experience I have actually lost far more time fiddling around with Arc and Arc_ecto than I would have had I just rolled my own solution explicitly uploading and deleting each file.

If you guys can get Bow as polished as Carrierwave and as well documented, I think there is a value proposition. Otherwise I think you should just use it internally.

In the interim I will get over my burn out and give it a shot one more time in Bow. At this point given the pain and time lost with Arc, I'm leaning more towards roll my own.

The one thing I really liked was deleting one version object deleted all versions. That was pretty slick :)

**Comment by smoku** [comment]
Please differentiate between arc, arc_ecto and ecto_arc. Rolling your own solution is a valid point, this is exactly what we did ourselves.

**Comment by ghost** [comment]
Yes I think it would be good as a learning experience. More tedious but I'd be a bit closer to the metal. Btw, is Bow a replacement for Arc or just what you folks are using yourself?

I'm not sure what the difference is between arc_ecto and ecto_arc. Only issue I couldn't get working was overriding filename being reflected in the db with arc_ecto.

Thank you for the feedback. At the moment I'm not able to get it working with Bow so I'll either settle with original file name or roll an implementation with ex_aws closer to the metal. 

The interaction with s3 is pretty smooth, I think Arc and Arc_ecto are about 99% there. Any plugin that simplifies the tedium of uploading, renaming and deleting images or files in general locally or to the cloud is a great time saver. So thanks for the speedy response, I hope my feedback was helpful.

**Comment by ghost** [comment]
If you need any help with documentation, please let me know. I can contribute about 5 hours a week.
